lib/flow/VaporField.cpp

```diff
@@ -0,0 +1,424 @@
+            assert( rv == 0 );
+            float ceilScalar = grid->GetValue( coords );
+            if( ceilScalar  == grid->GetMissingValue() )
+            {
+                scalar = 0.0f;
+                return 0;
+            }
+            float weight = (time - _timestamps[floorTS]) /
+                           (_timestamps[floorTS+1] - _timestamps[floorTS]);
+            scalar = glm::mix( floorScalar, ceilScalar, weight );
+        }
+    }
+
+    return 0;
+}
+
+
+bool
+VaporField::_isReady() const
+{
+    if( !_datamgr )         return false;
+    if( !_params )          return false;
+
+    return true;
+}
+
+
+void
+VaporField::AssignDataManager( VAPoR::DataMgr* dmgr )
+{
+    _datamgr = dmgr;
+
+    // Make a copy of the timestamps from the new data manager
+    const auto& timeCoords = dmgr->GetTimeCoordinates();
+    _timestamps.resize( timeCoords.size() );
+    for( size_t i = 0; i < timeCoords.size(); i++ )
+        _timestamps[i] = timeCoords[i];
+}
+
+
+void
+VaporField::UpdateParams( const VAPoR::FlowParams* p )
+{
+    _params = p;
+    // Update properties of this Field
+    IsSteady = p->GetIsSteady();
+    ScalarName = p->GetColorMapVariableName();
+    auto velNames = p->GetFieldVariableNames();
+    for( int i = 0; i < 3; i++ )
+        VelocityNames[i] = velNames.at(i);
+}
+
+
+int
+VaporField::LocateTimestamp( float time, size_t& floor ) const
+{
+    if( _timestamps.size() == 0 )
+        return TIME_ERROR;
+    if( _timestamps.size() == 1 )
+    {
+        if( _timestamps[0] != time )
+            return TIME_ERROR;
+        else
+        {
+            floor = 0;
+            return 0;
+        }
+    }
+
+    if( time < _timestamps.front() || time > _timestamps.back() )
+        return TIME_ERROR;
+    else
+    {
+        auto it = std::upper_bound( _timestamps.cbegin(), _timestamps.cend(), time );
+        floor = --it - _timestamps.cbegin();
+        return 0;
+    }
+}
+
+
+int
+VaporField::GetNumberOfTimesteps()
+{
+    return _timestamps.size();
+}
+
+
+int
+VaporField::_getAGrid( size_t               timestep,
+                       std::string&         varName,
+                       const VAPoR::Grid**  gridpp  )
+{
+    // First check if we have the requested grid in our cache.
+    // If it exists, return the grid directly.
+    std::vector<double>           extMin, extMax;
+    _params->GetBox()->GetExtents( extMin, extMax );
+    int refLevel  = _params->GetRefinementLevel();
+    int compLevel = _params->GetCompressionLevel();
+    std::string key = _paramsToString( timestep, varName, refLevel, compLevel, 
+                                       extMin, extMax );
+    if( _recentGrids.contains( key ) )
+    {
+        *gridpp = _recentGrids.find( key );
+        return 0;
+    }
+
+    // There's no such grid in our cache! Let's ask for it from the data manager,
+    // and then keep it in our cache!
+    VAPoR::Grid* grid = _datamgr->GetVariable( timestep, varName, refLevel, compLevel,
```

Comment:
GetVariable() will create a new Grid instance. This will also need to be freed. How is it being deleted?

Reply:
It is guaranteed to be unlocked and cleaned by the `GridWrapper` class, which itself is guaranteed to be cleaned up by the cache.